Fix fake kernel for the `out=...` variant of `unbind_copy`

StrongerXi opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #156643
* #156642

`unbind_copy(..., out=...)` returns None rather than the `out` argument
(see https://github.com/pytorch/pytorch/issues/130829#issuecomment-2283936222),
but the old fake kernel didn't account for that and caused an assertion
failure in `pushPyOutToStack`. This patch fixes that.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames

StrongerXi committed:
Update

[ghstack-poisoned]

StrongerXi committed:
Update

[ghstack-poisoned]

StrongerXi committed:
Update

[ghstack-poisoned]

StrongerXi commented on torch/_refs/__init__.py:
```diff
@@ -6631,7 +6633,7 @@ def select_scatter(x: TensorLikeType, src: TensorLikeType, dim: int, index: int)
 t_copy = _make_copy_from_view(aten.t)
 transpose_copy = _make_copy_from_view(aten.transpose)
-unbind_copy = _make_copy_from_view(aten.unbind)
+unbind_copy = _make_copy_from_view(aten.unbind, return_none_on_out_variant=True)
```
I thought about getting rid of this flag by making a special cased `@register_decomposition(aten.unbind_copy)` but got scared by all the scaffolding we have in `out_wrapper`: https://github.com/pytorch/pytorch/blob/5dd9652389ed7959a842323e4ce063f553710e47/torch/_prims_common/wrappers.py#L233-L405

StrongerXi replied:
And indeed they matter... Updated the PR to fix some newly exposed bugs.

zou3519 replied:
yeah I think that's fine

StrongerXi committed:
Update

[ghstack-poisoned]

StrongerXi committed:
Update on "Fix fake kernel for the `out=...` variant of `unbind_copy`"


`unbind_copy(..., out=...)` returns None rather than the `out` argument
(see https://github.com/pytorch/pytorch/issues/130829#issuecomment-2283936222),
but the old fake kernel didn't account for that and caused an assertion
failure in `pushPyOutToStack`. This patch fixes that.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

StrongerXi committed:
Update on "Fix fake kernel for the `out=...` variant of `unbind_copy`"


`unbind_copy(..., out=...)` returns None rather than the `out` argument
(see https://github.com/pytorch/pytorch/issues/130829#issuecomment-2283936222),
but the old fake kernel didn't account for that and caused an assertion
failure in `pushPyOutToStack`. This patch fixes that.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

zou3519 commented on torch/_dynamo/variables/torch.py:
```diff
@@ -1325,15 +1325,35 @@ def patched_fn(*args, **kwargs):
-            # guards. So save the shape now, and check later if it has
-            # changed. If it has, graph break.
-            fake_out_shape = kwargs["out"].proxy.node.meta["example_value"].shape
+
```
Can you explain a bit about why this changed?

StrongerXi replied:
Oops forgot.

Basically all the code here handles `torch.some_op(..., out=...)`, and in principle we only have 2 cases -- `out` is a tensor or a tuple/list of tensors, and we just need to check and graph break on contiguity/resize.

But the old way was a bit convoluted for historical reasons, and the special branch to handle `torch.some_op(..., out=...)` _returning_ `None` forgot to check for resize, and my change exposed this bug: https://github.com/pytorch/pytorch/blob/4515aaa7cc6fed3c9908586dab105b96c114d185/torch/_dynamo/variables/torch.py#L1389-L1408

So I just went and refactored the code dup and fixed the bug here.

zou3519 commented on torch/_dynamo/variables/torch.py:
```diff
@@ -1325,15 +1325,35 @@ def patched_fn(*args, **kwargs):
+                        all_tensors = False
+                        break
+                if all_tensors:
+                    saved_out_shapes = temp
```
why don't we save the shapes for a list of mixed Tensor and None? (maybe the answer is that we don't have an example for this)

StrongerXi replied:
> (maybe the answer is that we don't have an example for this)

Yeah, but I just tested and it errors in eager, so to fully make this compatible, I can raise an user error here (the old impl silently ignores the error, like this PR does now).

zou3519 replied:
sounds good

StrongerXi replied:
Actually I just kept it as a list of mixed Tensor and None, because otherwise it could be BC breaking in theory... https://github.com/pytorch/pytorch/blob/e15ea965a1e84029eb900b93f5776767a39fd91e/torch/_dynamo/variables/torch.py#L1324-L1327

zou3519 commented on torch/_dynamo/variables/torch.py:
```diff
@@ -1371,26 +1388,28 @@ def patched_fn(*args, **kwargs):
+                            # properly across dynamo/aot/inductor, just fall back.
+                            unimplemented("out variants with resizing on graph inputs")
+                        if not torch._prims_common.is_contiguous(fake_out):
+                            # It's difficult to handle strides correctly in functionalization
+                            # when calling an out= op with a non-contiguous out argument
+                            unimplemented(
+                                "out= op was called where output tensor was non-contiguous"
+                            )
```
to check, this path is new?

zou3519 replied:
interesting it seems like it was in the Tensor path but not this path

StrongerXi committed:
Update on "Fix fake kernel for the `out=...` variant of `unbind_copy`"


`unbind_copy(..., out=...)` returns None rather than the `out` argument
(see https://github.com/pytorch/pytorch/issues/130829#issuecomment-2283936222),
but the old fake kernel didn't account for that and caused an assertion
failure in `pushPyOutToStack`. This patch fixes that.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]
